src/custom-element.js

```diff
@@ -594,67 +601,102 @@ function createBaseCustomElementClass(win) {
     }
 
     /**
-     * If the element has a media attribute, evaluates the value as a media
-     * query and based on the result adds or removes the class
-     * `i-amphtml-hidden-by-media-query`. The class adds display:none to the
-     * element which in turn prevents any of the resource loading to happen for
-     * the element.
-     *
-     * This method is called by Resources and shouldn't be called by anyone
-     * else.
-     *
+     * Part of CE API. Called when a subscribed attribute has changed. See
+     * `observedAttributes` for the set of attributes.
+     * @param {string} name
      * @final
-     * @package
      */
-    applySizesAndMediaQuery() {
-      assertNotTemplate(this);
+    attributeChangedCallback(name) {
+      if (this.isConnected_ && MEDIA_ATTRS.includes(name)) {
+        this.initMediaAttrs_();
+      }
+    }
 
-      // Media query.
-      if (this.mediaQuery_ === undefined) {
-        this.mediaQuery_ = this.getAttribute('media') || null;
+    /** @private */
+    initMediaAttrs_() {
+      const hasMediaAttrs =
+        this.hasAttribute('media') ||
+        this.hasAttribute('sizes') ||
+        this.hasAttribute('heights');
+      const hadMediaAttrs = !!this.mediaQueryProps_;
+      const win = this.ownerDocument.defaultView;
+      if (hasMediaAttrs != hadMediaAttrs && win) {
+        if (hasMediaAttrs) {
+          this.mediaQueryProps_ = new MediaQueryProps(win, () =>
+            this.applyMediaAttrs_()
+          );
+        } else {
+          this.disposeMediaAttrs_();
+        }
       }
-      if (this.mediaQuery_) {
-        const {defaultView} = this.ownerDocument;
-        this.classList.toggle(
-          'i-amphtml-hidden-by-media-query',
-          !defaultView.matchMedia(this.mediaQuery_).matches
-        );
+      this.applyMediaAttrs_();
```

Comment:
I feel like this call is going to case every element to get `i-amphtml-hidden-by-media-query` class added, right? 

1. We always `initMediaAttrs_`
   - it only sets `mediaQueryProps_` if there's a media query prop
   - it always calls `applyMediaAttrs_`
2. `applyMediaAttrs_` detects that `!props` is true
   - we add `i-amphtml-hidden-by-media-query`

Reply:
`applyMediaAttrs_` always removes `i-amphtml-hidden-by-media-query` when `!props`. I.e. it cleans up. But a similar point here as in [this thread](https://github.com/ampproject/amphtml/pull/31948#discussion_r557003634) - if we ignore attribute changes, that I can simplify this code further.